Generate code coverage with Jacoco

TurpIF opened the pull request.

Amokrane commented:
@TurpIF Could you introduce a few code smells and defects on purpose to see how it performs?
FYI, I have compared both `detekt` and `SonarQube`'s Kotlin rules and `detekt` is much more advanced. For instance, `SonarQube` does not have any rules for coroutines whereas `detekt` does (https://detekt.github.io/detekt/coroutines.html). Same thing for performance related rules which are more specific to Kotlin. It would be good to have both (detekt has a SonarQube integration). That said, we can do it later.

TurpIF commented:
> @TurpIF Could you introduce a few code smells and defects on purpose to see how it performs?
> FYI, I have compared both `detekt` and `SonarQube`'s Kotlin rules and `detekt` is much more advanced. For instance, `SonarQube` does not have any rules for coroutines whereas `detekt` does (https://detekt.github.io/detekt/coroutines.html). Same thing for performance related rules which are more specific to Kotlin. It would be good to have both (detekt has a SonarQube integration). That said, we can do it later.

Done: https://sonarcloud.io/code?id=TurpIF_android-coverage-test&pullRequest=3&selected=TurpIF_android-coverage-test%3Alib%2Fsrc%2Fmain%2Fjava%2Ffr%2Fpturpin%2FCodeSmells.kt

Amokrane commented:
Never mind, clicked approve accidentally :)

TurpIF committed:
Add dummy content to show the effect of the coverage report in PR

When doing a new PR, it is expected that the Sonar bot will comment and
shows the coverage report on the new code of the PR.
On 3 methods, 2 are covered by tests (one by Java tests, one by Android
tests), then we can expect the coverage ratio on the new code to be
approximately 66%.

TurpIF committed:
Generate XML and HTML coverage reports when running android tests

The Jacoco task from the Android Gradle plugin is completely custom. It
generates automatically reports but it does not support cross-modules
coverage.
As the android tests are in a dedicated module, by default, the Android
Jacoco will not gather the coverage of the lib module. The workaround is
to make the sources of the lib belong to the sources of the test module.
This mess up the IDE, so this workaround is only applied when Gradle is
invoked with a dedicated parameter.

TurpIF committed:
Generate XML and HTMl coverage reports when running JVM tests

The Jacoco task from Gradle generates execution data but does not
automatically generate reports on Android modules.
This reports generation task is added on all subprojects using the
jacoco plugin.

TurpIF committed:
Send coverage reports to Sonar

To speed up the execution, JVM and Android tests are run in parallel on
the CI. So coverage reports are gathered by each job and then given to
another CI job who will send them to Sonar.
Sonar would then have the responsibility to merge the reports.
